clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -578,11 +578,6 @@ class DiagDeviceFunction : public RecursiveASTVisitor<DiagDeviceFunction> {
             << Sema::KernelCallRecursiveFunction;
       }
 
-      if (const CXXMethodDecl *Method = dyn_cast<CXXMethodDecl>(Callee))
```

Comment:
This should be guarded by an option, because SYCL core spec doesn't allow to use virtual functions in device code. I think you can implement something similar to existing `-fsycl-allow-func-ptr` flag